src/Features/Core/Portable/Diagnostics/DiagnosticResultSerializer.cs

```diff
@@ -68,82 +70,103 @@ internal static class DiagnosticResultSerializer
                     {
-                        syntaxLocalMap = ReadDiagnosticDataMap(reader, diagnosticDataSerializer, project, cancellationToken);
-                        semanticLocalMap = ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty;
+                        ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>? syntaxLocalMap, semanticLocalMap;
+                        if (documentAnalysisScope.Kind == AnalysisKind.Syntax)
+                        {
+                            if (!TryReadDiagnosticDataMap(reader, diagnosticDataSerializer, project, cancellationToken, out syntaxLocalMap))
+                            {
+                                return false;
+                            }
+
+                            semanticLocalMap = ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty;
+                        }
+                        else
+                        {
+                            Debug.Assert(documentAnalysisScope.Kind == AnalysisKind.Semantic);
+                            if (!TryReadDiagnosticDataMap(reader, diagnosticDataSerializer, project, cancellationToken, out semanticLocalMap))
+                            {
+                                return false;
+                            }
+
+                            syntaxLocalMap = ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty;
+                        }
+
+                        analysisResult = DiagnosticAnalysisResult.Create(
+                            project,
+                            version,
+                            syntaxLocalMap,
+                            semanticLocalMap,
+                            nonLocalMap: ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty,
+                            others: ImmutableArray<DiagnosticData>.Empty,
+                            documentIds);
                     }
                     else
                     {
-                        Debug.Assert(documentAnalysisScope.Kind == AnalysisKind.Semantic);
-                        semanticLocalMap = ReadDiagnosticDataMap(reader, diagnosticDataSerializer, project, cancellationToken);
-                        syntaxLocalMap = ImmutableDictionary<DocumentId, ImmutableArray<DiagnosticData>>.Empty;
+                        if (!TryReadDiagnosticDataMap(reader, diagnosticDataSerializer, project, cancellationToken, out var syntaxLocalMap) ||
+                            !TryReadDiagnosticDataMap(reader, diagnosticDataSerializer, project, cancellationToken, out var semanticLocalMap) ||
+                            !TryReadDiagnosticDataMap(reader, diagnosticDataSerializer, project, cancellationToken, out var nonLocalMap) ||
+                            !diagnosticDataSerializer.TryReadDiagnosticData(reader, project, document: null, cancellationToken, out var others))
+                        {
+                            return false;
```

Comment:
when does this happen.  seems bad.

Reply:
Down the stack there are checks for version mismatches, and also there was an existing try/catch for reading diagnostic data. Those code paths now return false instead of silently skipping reading of expected data.

Reply:
What is the try/catch catching? Should that rather keep the exception propagating?

---
In reply to: [466700707](https://github.com/dotnet/roslyn/pull/46597#discussion_r466700707) [](ancestors = 466700707)

Reply:
The version mismatch should probably also throw (ExceptionUtilities.UnexpectedValue).

---
In reply to: [468051167](https://github.com/dotnet/roslyn/pull/46597#discussion_r468051167) [](ancestors = 468051167,466700707)

Reply:
So you feel it is better to let the exception go unhandled and crash VS rather then log an NFW? It is not clear to me why the original code was bailing out on bunch of version checks and had a try catch for deserialization, but I still feel the diagnostic deserialization part should not bring down VS, similar to the way any exception inside CompilationWithAnalyzers for exception in analyzer execution does not bring down VS.

Reply:
It should not bring down VS. There should be try-catch somewhere higher on the stack. In OOP, this would be already handled on the remoting boundary. In devenv, the diagnostic analyzer runner should have try-catch somewhere.

Reply:
Ah, that makes sense to me. Let me find a better place to add the handler and update the PR.

Reply:
See https://github.com/dotnet/roslyn/pull/46597#discussion_r490296477, I feel much more comfortable making the deserialize API safer with `bool TryDeserializeXXX` approach here.